### PR TITLE
minor: loop-that-never-loops

### LINKED ISSUE #13921: This loop never loops
https://github.com/rust-lang/rust-analyzer/blob/368e0bb32f1178cf162c2ce5f7e10b7ae211eb26/crates/hir-def/src/attr.rs#L254-L266

source: clippy.

After (locally) addressing #13919 and #13920 I see a bunch more useful clippy lints. Squashing clippy lints you don't necessarily care about can in many cases provide visibility of lints you probably do care about.

### PR DESCRIPTION
**Pull request by danieleades**
closes #13921

### CODE REVIEW

**Comment by lnicola** [comment]
I think I'd prefer the break-with-value version, but whatever.

@bors r+

**Comment by HKalbasi** [comment]
> I think I'd prefer the break-with-value version, but whatever.

Why? I think `.find` says the intention behind the for loop more clearly. If it didn't mutate `idx` and instead used `.enumerate` it was even better.

**Comment by lnicola** [comment]
Personal preference, I find imperative code easier to read. Good point about `enumerate` though.

**Comment by lnicola** [comment]
@bors r-

:grinning:

**Comment by Veykril** [comment]
I have to agree, I prefer the imperative version here as well :)

+1 to using enumerate though, if that is used the iterator approach should look fine to me as well.

**Comment by danieleades** [comment]
i would love to use `enumerate`, but the index is only conditionally incremented. `idx` is incremented iff `attrs.is_cfg_enabled(cfg_options)`

**Comment by HKalbasi** [comment]
You can filter the iterator per the first condition, then enumerate, then find.

**Comment by danieleades** [comment]
> You can filter the iterator per the first condition, then enumerate, then find.

yep, that's the approach i've taken.

I *still* haven't used `enumerate`, because i need the loop index to be a `u32`, not a `usize`. The intent is the same, and I avoid a `try_into().unwrap()` call.

File: crates/hir-def/src/attr.rs [hunk 1]
```diff
@@ -251,19 +251,18 @@ impl Attrs {
...
-                        }
-                    }
+
+                let Some(variant) = enum_.variants.clone().filter(|variant| {
+                    let attrs = item_tree.attrs(db, krate, (*variant).into());
+                    attrs.is_cfg_enabled(cfg_options)
+                })
+                .zip(0u32..)
+                .find(|(_variant, idx)| it.local_id == Idx::from_raw(RawIdx::from(*idx)))
+                .map(|(variant, _idx)| variant)
```
Comment (danieleades) [inline crates/hir-def/src/attr.rs:261]: i can collapse `find` and `map` into a `find_map`, but it's actually more concise to do it this way. I'm not married to either approaches

Comment (HKalbasi) [reply crates/hir-def/src/attr.rs:261]: You can also ignore the index in the let else pattern.

Comment (danieleades) [reply crates/hir-def/src/attr.rs:261]: I certainly could

**Comment by danieleades** [comment]
@lnicola anything left to do on this one? I note it's marked as 'waiting on author'

**Comment by HKalbasi** [comment]
I don't have r+ access on this repo, but this looks good to me.

About waiting on author label, you can change it with this command:
@rustbot ready

**Comment by lnicola** [comment]
> I still haven't used enumerate, because i need the loop index to be a u32, not a usize.

We don't mind casting between those, no need for `try_into`s and I'm still not a fan of that `zip`. But, anyway, see how long it can take to make a simple style change?

@bors r+

**Comment by matklad** [comment]
>Personal preference, I find imperative code easier to read. Good point about enumerate though.

Or rather, it's the official codebase style

https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/style.md#functional-combinators


Nice branch name!

**Comment by danieleades** [comment]
i personally don't have a dog in the imperative vs functional style fight, but using (ab)using the labelled break feature of a loop that never actually looped was, i felt, 'surprising' enough to obscure the intent.
It would certainly be possible to write this in a way that was both imperative *and* avoided abusing a `loop`, though i think everyone in this thread has probably put too much energy into this snippet already ;)

**Comment by matklad** [comment]
I think these days we can break out of blocks?

**Comment by Veykril** [comment]
We can now, the loop was used because label break value was not supported on stable yet. In a similar fashion we are using `(|| { ... })()` for stable try blocks in some places :)

**Pull request merged**